build_defs/build_defs.bzl

```diff
@@ -39,17 +44,60 @@ def arcs_cc_schema(name, src, out = None):
         file_extension = ".h",
     )
 
-def arcs_kt_schema(name, src, out = None):
+def _wasm_name(name):
+    return name + "_wasm"
+
+def _wasm_deps(deps):
+    return [_wasm_name(dep) for dep in deps]
+
+def _jvm_and_wasm_library(name, srcs, deps):
+    kt_android_library(
+        name = name,
+        srcs = srcs,
+        deps = deps
+    )
+
+    kt_wasm_library(
+        name = _wasm_name(name),
+        srcs = srcs,
+        deps = _wasm_deps(deps)
+    )
+
+def arcs_kt_schema(name, src, out = None, deps = [], wasm_deps = []):
     """Generates a Kotlin file for the given .arcs schema file."""
+    out = out or _output_name(src, ".kt")
     _run_schema2pkg(
-        name = name,
+        name = name + "_genrule",
         src = src,
         out = out,
         language_name = "Kotlin",
         language_flag = "--kotlin",
         file_extension = ".kt",
     )
 
+    return out
+
+def arcs_manifest(name, srcs, deps = []):
+    """Generates Kotlin+Wasm Library targets from .arcs file translated to Kotlin."""
+    # TODO: when converted to rule()+provider() verify deps are arcs_manifest rules
+    outs = [arcs_kt_schema(_output_name(src), src, deps) for src in srcs]
+    deps = ["//src/wasm/kotlin:arcs"] + deps
+
+    _jvm_and_wasm_library(
+        name,
+        outs,
+        deps
+    )
+
+def arcs_kt_particle(name, srcs, deps = []):
+    """Declares kotlin library/wasm targets for Kotlin particle sources."""
+    _jvm_and_wasm_library(name, srcs, deps)
+
+def arcs_deployable(name, deps):
```

Comment:
How about `arcs_kt_wasm_binary` instead? I don't instantly equate "deployable" with "wasm".

Also, this looks like it's only for Kotlin, so it seems like it should have "kt" in the name somewhere? And at that point, it's essentially the same rule as `arcs_kt_particle`. We could just delete this rule and save developers having to add two sets of rules (e.g. `service_particle_lib` and `service_particle` from the examples below).

I'd recommend just boiling this all down to: `arcs_kt_wasm_particle`, `arcs_kt_jvm_particle`, `arcs_cc_wasm_particle`.

Or we could be less verbose and do something like: `arcs_kt_particle` (implies wasm), `arcs_android_particle` (implies jvm, kt), `arcs_cc_particle` (implies wasm).

Reply:
There's a difference between libraries and particles. arcs_kt_particle is analogous to java_library/kotlin klib, whereas arcs_deployable is analogous to java_binary/kt_wasm_binary. They do difference things. 

You can't delete the rule if you plan to have inter-particle dependencies (particle1 deps on particle2), you also can't delete the rule if you want to support multiple-particles in a single wasm file. 

If arcs_kt_particle is to support both wasm and JVM, then you have a dichotomy, because the JVM version would produce a library, which ultimately is bundled into an Android APK, but before that is just a JAR, but the WASM version would actually be producing not klibs, but .wasm+.wasm.js .  If the source language is the same, and the only thing that differs is the compilation target output (bytecode vs wasm vs x86 etc), I think having a separate JVM vs WASM rule for Kotlin is bad for code sharing, because with Particle Accelerator, the same particle code can run without change, in either environment, yet this would seem to dictate explicit 2x rules for everything. 

I'm fine with changing it to arcs_kt_wasm_binary, because APK-builtin JVM code can't be 'deployed' (yet!) but note that eventually the need for a 'wasm specific' rule might not be needed once bazel/blaze toolchain support is available, so how we deal with that will be another issue. 

I'm just trying to arrive a set of build rules for Kotlin particles that are maximally flexible, and to me, WASM is merely a platform cross compilation choice, no different than cross compiling between x86, arm, and mips.

Reply:
Added arcs_kt_wasm_binary